### PR TITLE
[API] Paginate list pods/crds in periodic monitoring and cleanup

### PR DESCRIPTION
**Pull request by alonmr**
We are seeing that listing pods in cleanup/monitoring cycle is consuming to much memory which causes the API memory consumption to increase over time and eventually kill the chief API container. Paginating the requests to k8s helps reducing the memory peaks which significantly improves the API consumption over time.

https://iguazio.atlassian.net/browse/ML-6317

### CODE REVIEW

File: server/api/utils/singletons/k8s.py [hunk 1]
```diff
@@ -125,6 +125,115 @@ def list_pods(self, namespace=None, selector="", states=None):
...
+                    label_selector=selector,
+                    watch=False,
+                    limit=200,
+                )
+            else:
+                pods_list = self.v1api.list_namespaced_pod(
+                    self.resolve_namespace(namespace),
+                    label_selector=selector,
+                    watch=False,
+                    limit=200,
```
Comment (liranbg) [inline server/api/utils/singletons/k8s.py]: Make configurable

File: server/api/utils/singletons/k8s.py [hunk 2]
```diff
@@ -125,6 +125,115 @@ def list_pods(self, namespace=None, selector="", states=None):
...
+        _continue = ""
+        logger.debug(
+            "Listing namespaced pods with pagination",
+            selector=selector,
+            namespace=namespace,
+            states=states,
+        )
+        while True:
+            if not _continue:
+                pods_list = self.v1api.list_namespaced_pod(
```
Comment (liranbg) [inline server/api/utils/singletons/k8s.py]: DRY

File: server/api/utils/singletons/k8s.py [hunk 3]
```diff
@@ -125,6 +125,115 @@ def list_pods(self, namespace=None, selector="", states=None):
...
+                    watch=False,
+                    limit=200,
+                    _continue=_continue,
+                )
+
+            for item in pods_list.items:
+                if not states or item.status.phase in states:
+                    yield item
+
+            _continue = pods_list._metadata._continue
```
Comment (liranbg) [inline server/api/utils/singletons/k8s.py]: No better way to resolve the continue?

**Review by liranbg — COMMENTED** [review]
Very well. Minor stuff

File: server/api/utils/singletons/k8s.py (hunk 3 above)
Comment (alonmr) [reply server/api/utils/singletons/k8s.py]: metadata has a public getter but continue doesn't
also still checking how this works for crds

File: server/api/utils/singletons/k8s.py [hunk 4]
```diff
@@ -125,6 +125,99 @@ def list_pods(self, namespace=None, selector="", states=None):
                 items.append(i)
         return items
 
+    @raise_for_status_code
+    def list_pods_paginated(self, namespace=None, selector="", states=None):
```
Comment (quaark) [inline server/api/utils/singletons/k8s.py:129]: Feels to me like a lot of these 2 functions can be generalized into a `list_paginated` method that receives a list method callback, and can be later reused by `list_pods_paginated`, `list_crds_paginated` and any other resource we want.

**Review by quaark — COMMENTED** [review]
Looks great!
I have a suggestion, I feel like we can save some code dup

File: server/api/utils/singletons/k8s.py (hunk 4 above)
Comment (alonmr) [reply server/api/utils/singletons/k8s.py:129]: I'm not sure I agree. There are differences with how to parse the result which needs to be part of the callback and it turns out I'm extracting 90% of the method to the callback. If you have some other Idea lets discuss offline but in general I think that the common part is not big enough to make this refactor.

**Pull request merged**